File: 程序员面试金典/2.1移除重复节点/removeDuplicateNodes.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeDuplicateNodes(self, head: ListNode) -> ListNode:
        seen = set()
        new_head = ListNode(0)
        new_head.next = head
        p = new_head
        while p.next is not None:
            if p.next.val in seen:
                p.next = p.next.next
            else:
                seen.add(p.next.val)
                p = p.next
        return new_head.next

File: 程序员面试金典/2.1/removeDuplicateNodes.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 使用散列表，时间复杂度O(n)，空间复杂度O(n)
class Solution {
    public ListNode removeDuplicateNodes(ListNode head) {
        Set<Integer> seen = new HashSet<>();
        ListNode prev = null, p = head;
        while (p != null) {
            if (seen.contains(p.val))
                prev.next = p.next;
            else {
                seen.add(p.val);
                prev = p;
            }
            p = p.next;
        }
        return head;
    }
}
